salt.crypt broken in develop

using the lxc runner, which calls config.seed, wich generate keys fails
```
[INFO    ] Executing command "lxc-attach --clear-env --set-var PATH=/bin:/usr/bin:/sbin:/usr/sbin:/opt/bin:/usr/local/bin:/usr/local/sbin  -n foobar24 -- test -e '/lxc.initial_seed'" in directory '/root'
[DEBUG   ] retcode: 1
[DEBUG   ] LazyLoaded seed.mkconfig
[WARNING ] The minion function caused an exception
Traceback (most recent call last):
  File "/salt-venv/salt/src/salt/salt/minion.py", line 1005, in _thread_return
    return_data = func(*args, **kwargs)
  File "/salt-venv/salt/src/salt/salt/modules/lxc.py", line 1487, in init
    bootstrap_args=bootstrap_args)
  File "/salt-venv/salt/src/salt/salt/modules/lxc.py", line 2891, in bootstrap
    pub_key=pub_key, priv_key=priv_key)
  File "/salt-venv/salt/src/salt/salt/modules/seed.py", line 220, in mkconfig
    salt.crypt.gen_keys(tmp, 'minion', 2048)
  File "/salt-venv/salt/src/salt/salt/crypt.py", line 87, in gen_keys
    gen = RSA.generate(bits=keysize, e=65537)
  File "/usr/lib/python2.7/dist-packages/Crypto/PublicKey/RSA.py", line 508, in generate
    obj = _RSA.generate_py(bits, rf, progress_func, e)    # TODO: Don't use legacy _RSA module
  File "/usr/lib/python2.7/dist-packages/Crypto/PublicKey/_RSA.py", line 50, in generate_py
    p = pubkey.getStrongPrime(bits>>1, obj.e, 1e-12, randfunc)
  File "/usr/lib/python2.7/dist-packages/Crypto/Util/number.py", line 265, in getStrongPrime
    randfunc)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 202, in read
    return self._singleton.read(bytes)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 178, in read
    return _UserFriendlyRNG.read(self, bytes)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 132, in read
    retval = self._fa.random_data(N)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/Fortuna/FortunaAccumulator.py", line 148, in random_data
    return self.generator.pseudo_random_data(bytes)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/Fortuna/FortunaGenerator.py", line 93, in pseudo_random_data
    for i in xrange(num_full_blocks):
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 202, in read
    return self._singleton.read(bytes)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 178, in read
    return _UserFriendlyRNG.read(self, bytes)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 132, in read
    retval = self._fa.random_data(N)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/Fortuna/FortunaAccumulator.py", line 148, in random_data
    return self.generator.pseudo_random_data(bytes)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/Fortuna/FortunaGenerator.py", line 93, in pseudo_random_data
    for i in xrange(num_full_blocks):
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 202, in read
    return self._singleton.read(bytes)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 178, in read
    return _UserFriendlyRNG.read(self, bytes)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 137, in read
    self._check_pid()
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 153, in _check_pid
    raise AssertionError("PID check failed. RNG must be re-initialized after fork(). Hint: Try Random.atfork()")
AssertionError: PID check failed. RNG must be re-initialized after fork(). Hint: Try Random.atfork()

```

--- a/salt/cloud/__init__.py
+++ b/salt/cloud/__init__.py
@@ -32,6 +32,7 @@ import salt.loader
 import salt.utils
 import salt.utils.cloud
 import salt.syspaths
+from salt.utils import reinit_crypto
 from salt.utils import context
 from salt.ext.six import string_types
 from salt.template import compile_template
@@ -2236,8 +2237,7 @@ def run_parallel_map_providers_query(data, queue=None):
     This function will be called from another process when building the
     providers map.
     '''
-    if 'Crypto.Random' in sys.modules:
-        Crypto.Random.atfork()
+    reinit_crypto()
 
     cloud = Cloud(data['opts'])
     try:

--- a/salt/crypt.py
+++ b/salt/crypt.py
@@ -23,7 +23,8 @@ try:
     from Crypto.Hash import SHA
     from Crypto.PublicKey import RSA
     from Crypto.Signature import PKCS1_v1_5
-    import Crypto.Random
+    # let this be imported, if possible
+    import Crypto.Random  # pylint: disable=W0611
 except ImportError:
     # No need for crypt in local mode
     pass

--- a/salt/crypt.py
+++ b/salt/crypt.py
@@ -361,8 +361,7 @@ class AsyncAuth(object):
 
         self.io_loop = io_loop or tornado.ioloop.IOLoop.current()
 
-        if 'Crypto.Random' in sys.modules:
-            Crypto.Random.atfork()
+        salt.utils.reinit_crypto()
         key = self.__key(self.opts)
         # TODO: if we already have creds for this key, lets just re-use
         if key in AsyncAuth.creds_map:

--- a/salt/master.py
+++ b/salt/master.py
@@ -749,8 +749,7 @@ class MWorker(multiprocessing.Process):
             self.key,
             )
         self.aes_funcs = AESFuncs(self.opts)
-        if 'Crypto.Random' in sys.modules:
-            Crypto.Random.atfork()
+        salt.utils.reinit_crypto()
         self.__bind()
 
 

--- a/salt/master.py
+++ b/salt/master.py
@@ -18,10 +18,6 @@ import multiprocessing
 # Import third party libs
 import zmq
 from Crypto.PublicKey import RSA
-try:
-    import Crypto.Random
-except ImportError:
-    pass
 # pylint: disable=import-error,no-name-in-module,redefined-builtin
 import salt.ext.six as six
 from salt.ext.six.moves import range

--- a/salt/minion.py
+++ b/salt/minion.py
@@ -23,6 +23,7 @@ from stat import S_IMODE
 # pylint: disable=import-error,no-name-in-module,redefined-builtin
 import salt.ext.six as six
 from salt.ext.six.moves import range
+from salt.utils import reinit_crypto
 # pylint: enable=no-name-in-module,redefined-builtin
 
 # Import third party libs
@@ -653,8 +654,10 @@ class Minion(MinionBase):
                 log.debug('Starting {0} proxy.'.format(p))
                 pid = os.fork()
                 if pid > 0:
+                    reinit_crypto()
                     continue
                 else:
+                    reinit_crypto()
                     proxyminion = ProxyMinion(self.opts)
                     proxyminion.start(self.opts['pillar']['proxy'][p])
                     self.clean_die(signal.SIGTERM, None)

--- a/salt/modules/inspectlib/collector.py
+++ b/salt/modules/inspectlib/collector.py
@@ -25,6 +25,7 @@ from subprocess import Popen, PIPE, STDOUT
 from salt.modules.inspectlib.dbhandle import DBHandle
 from salt.modules.inspectlib.exceptions import (InspectorSnapshotException)
 from salt.utils import fsutils
+from salt.utils import reinit_crypto
 
 
 class Inspector(object):
@@ -346,7 +347,10 @@ if __name__ == '__main__':
     # Double-fork stuff
     try:
         if os.fork() > 0:
+            reinit_crypto()
             sys.exit(0)
+        else:
+            reinit_crypto()
     except OSError as ex:
         sys.exit(1)
 
@@ -356,6 +360,7 @@ if __name__ == '__main__':
     try:
         pid = os.fork()
         if pid > 0:
+            reinit_crypto()
             fpid = open(pidfile, "w")
             fpid.write("{0}\n".format(pid))
             fpid.close()
@@ -363,4 +368,5 @@ if __name__ == '__main__':
     except OSError as ex:
         sys.exit(1)
 
+    reinit_crypto()
     main(dbfile, pidfile, mode)

--- a/salt/utils/__init__.py
+++ b/salt/utils/__init__.py
@@ -347,10 +347,8 @@ def reinit_crypto():
         child processes after using os.fork()
 
     '''
-    ret = None
     if HAS_CRYPTO:
-        ret = Crypto.Random.atfork()
-    return ret
+        Crypto.Random.atfork()
 
 
 def daemonize(redirect_out=True):

--- a/salt/utils/__init__.py
+++ b/salt/utils/__init__.py
@@ -43,6 +43,13 @@ from stat import S_IMODE
 # pylint: enable=import-error,redefined-builtin
 
 # Try to load pwd, fallback to getpass if unsuccessful
+# Import 3rd-party libs
+try:
+    import Crypto.Random
+    HAS_CRYPTO = True
+except ImportError:
+    HAS_CRYPTO = False
+
 try:
     import pwd
 except ImportError:
@@ -330,6 +337,22 @@ def get_specific_user():
     return user
 
 
+def reinit_crypto():
+    '''
+    When a fork arrises, pycrypto needs to reinit
+    From its doc::
+
+        Caveat: For the random number generator to work correctly,
+        you must call Random.atfork() in both the parent and
+        child processes after using os.fork()
+
+    '''
+    ret = None
+    if HAS_CRYPTO:
+        ret = Crypto.Random.atfork()
+    return ret
+
+
 def daemonize(redirect_out=True):
     '''
     Daemonize a process
@@ -338,6 +361,7 @@ def daemonize(redirect_out=True):
         pid = os.fork()
         if pid > 0:
             # exit first parent
+            reinit_crypto()
             sys.exit(salt.defaults.exitcodes.EX_OK)
     except OSError as exc:
         log.error(
@@ -355,6 +379,7 @@ def daemonize(redirect_out=True):
     try:
         pid = os.fork()
         if pid > 0:
+            reinit_crypto()
             sys.exit(salt.defaults.exitcodes.EX_OK)
     except OSError as exc:
         log.error(
@@ -363,6 +388,8 @@ def daemonize(redirect_out=True):
             )
         )
         sys.exit(salt.defaults.exitcodes.EX_GENERIC)
+
+    reinit_crypto()
 
     # A normal daemonization redirects the process output to /dev/null.
     # Unfortunately when a python multiprocess is called the output is

--- a/salt/utils/vt.py
+++ b/salt/utils/vt.py
@@ -46,6 +46,7 @@ else:
     import resource
 
 # Import salt libs
+from salt.utils import reinit_crypto
 from salt.ext.six import string_types
 from salt.log.setup import LOG_LEVELS
 
@@ -486,6 +487,7 @@ class Terminal(object):
                 # Close parent FDs
                 os.close(stdout_parent_fd)
                 os.close(stderr_parent_fd)
+                reinit_crypto()
 
                 # ----- Make STDOUT the controlling PTY --------------------->
                 child_name = os.ttyname(stdout_child_fd)
@@ -546,6 +548,7 @@ class Terminal(object):
                 # <---- Duplicate Descriptors --------------------------------
             else:
                 # Parent. Close Child PTY's
+                reinit_crypto()
                 os.close(stdout_child_fd)
                 os.close(stderr_child_fd)
 

